core/src/apps/webauthn/credential.py

```diff
@@ -0,0 +1,313 @@
+_CRED_ID_USER_DISPLAY_NAME = const(0x05)
+_CRED_ID_CREATION_TIME = const(0x06)
+_CRED_ID_HMAC_SECRET = const(0x07)
+
+# Key paths
+_U2F_KEY_PATH = const(0x80553246)
+
+
+class Credential:
+    def __init__(self) -> None:
+        self.id = b""  # type: bytes
+        self.rp_id = ""  # type: str
+        self.rp_id_hash = b""  # type: bytes
+        self.user_id = None  # type: Optional[bytes]
+
+    def app_name(self) -> str:
+        return ""
+
+    def account_name(self) -> Optional[str]:
+        return None
+
+    def private_key(self) -> bytes:
+        return b""
+
+    def hmac_secret_key(self) -> Optional[bytes]:
+        return None
+
+    @staticmethod
+    def from_bytes(data: bytes, rp_id_hash: bytes) -> Optional["Credential"]:
+        cred = Fido2Credential.from_cred_id(
+            data, rp_id_hash
+        )  # type: Optional[Credential]
+        if cred is None:
+            cred = U2fCredential.from_key_handle(data, rp_id_hash)
+        return cred
+
+
+# SLIP-0022: FIDO2 credential ID format for HD wallets
+class Fido2Credential(Credential):
+    def __init__(self) -> None:
+        super().__init__()
+        self.rp_name = None  # type: Optional[str]
+        self.user_name = None  # type: Optional[str]
+        self.user_display_name = None  # type: Optional[str]
+        self._creation_time = 0  # type: int
+        self.hmac_secret = False  # type: bool
+
+    def __lt__(self, other: Credential) -> bool:
+        # Sort FIDO2 credentials newest first amongst each other.
+        if isinstance(other, Fido2Credential):
+            return self._creation_time > other._creation_time
+
+        # Sort FIDO2 credentials before U2F credentials.
+        return True
+
+    def generate_id(self) -> None:
+        from apps.common import seed
+
+        self._creation_time = storage.device.next_u2f_counter() or 0
+
+        data = cbor.encode(
+            {
+                key: value
+                for key, value in (
+                    (_CRED_ID_RP_ID, self.rp_id),
+                    (_CRED_ID_RP_NAME, self.rp_name),
+                    (_CRED_ID_USER_ID, self.user_id),
+                    (_CRED_ID_USER_NAME, self.user_name),
+                    (_CRED_ID_USER_DISPLAY_NAME, self.user_display_name),
+                    (_CRED_ID_CREATION_TIME, self._creation_time),
+                    (_CRED_ID_HMAC_SECRET, self.hmac_secret),
+                )
+                if value
+            }
+        )
+        key = seed.derive_slip21_node_without_passphrase(
+            [b"SLIP-0022", _CRED_ID_VERSION, b"Encryption key"]
+        ).key()
+        iv = random.bytes(12)
+        ctx = chacha20poly1305(key, iv)
+        ctx.auth(self.rp_id_hash)
+        ciphertext = ctx.encrypt(data)
+        tag = ctx.finish()
+        self.id = _CRED_ID_VERSION + iv + ciphertext + tag
+
+    @staticmethod
+    def from_cred_id(cred_id: bytes, rp_id_hash: bytes) -> Optional["Fido2Credential"]:
+        from apps.common import seed
```

Comment:
What is the reason that you used inside function import?

Reply:
This was done in the original U2F implementation of msg_register_sign() and msg_authenticate_genkey() so I followed suit. I assume the reason was to limit the visibility of the seed module. @jpochyla will know more.

Reply:
@jpochyla told me he is OK with putting these imports outside of the functions. The reason why they were originally in-function is unknown. There may have been some issues with unimporting modules, but that's only speculation. The visibility may be a relevant issue, but we don't care about it anywhere else, so it's questionable whether we should care about it here. On the other hand, perhaps we should care about it elsewhere ...